### PR TITLE
[router] RouterExceptionAndTrackingUtils allocation reduction

### PR DESCRIPTION
**Pull request by FelixGV**
Removed all Optionals from RouterExceptionAndTrackingUtils.

Miscellaneous:

- Tweaked flaky unit test RouterRequestThrottlingTest.

- Tweaked the main README page.

N.B.: This is another carve out from PR #1367, in order to make that other PR smaller.

## How was this PR tested?
CI.

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-common/src/test/java/com/linkedin/venice/meta/TestZKStore.java [hunk 1]
```diff
@@ -395,34 +395,6 @@ public void invalidStoreNameThrows() {
         1);
   }
 
-  /**
-   * We're not relying on push ID uniqueness.  We can reenable this test if we start relying on (and enforcing) push ID uniqueness
-   */
-  @Test(groups = { "flaky" })
-  public void cannotAddDifferentVersionsWithSamePushId() {
```
Comment (KaiSernLim) [inline internal/venice-common/src/test/java/com/linkedin/venice/meta/TestZKStore.java:402]: super nit: I think the removal of this unit test should go in the PR description.

File: services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java [hunk 2]
```diff
@@ -23,14 +24,27 @@
...
+
+    FailureType() {
+      this.reportUnhealthy = true;
+      this.statsRecorder = (stats, store) -> {}; // No op
+    }
+
+    FailureType(BiConsumer<AggRouterHttpRequestStats, String> statsRecorder) {
+      this.reportUnhealthy = false;
+      this.statsRecorder = Objects.requireNonNull(statsRecorder);
+    }
```
Comment (KaiSernLim) [inline services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java:47]: Very interesting. This makes the below `switch` statement so unnecessary.

File: services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java [hunk 3]
```diff
@@ -72,7 +74,12 @@ public static Object[][] requestType() {
     return new Object[][] { { RequestType.MULTI_GET }, { RequestType.COMPUTE } };
   }
 
-  @Test(timeOut = 30000, dataProvider = "multiGet_compute")
+  /** A simple Runnable-like interface just to reduce boilerplate in the test... */
+  private interface ScatterCall {
+    void run() throws RouterException;
+  }
```
Comment (KaiSernLim) [inline services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java:80]: I should be doing this more often...

File: services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java [hunk 4]
```diff
@@ -72,7 +74,12 @@ public static Object[][] requestType() {
...
+
+  @Test(timeOut = 30000, dataProvider = "multiGet_compute", invocationCount = 100)
```
Comment (KaiSernLim) [inline services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java:82]: Is the `invocationCount = 100` intentional?

**Review by KaiSernLim — APPROVED** [review]

File: services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java (hunk 4 above)
Comment (FelixGV) [reply services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java:82]: Oh, no it isn't, thanks for catching that. I only did this to make sure my de-flaking worked.

File: services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java (hunk 2 above)
Comment (FelixGV) [reply services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java:47]: Yep, adding behavior to enums can be very useful...!

File: services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java (hunk 3 above)
Comment (FelixGV) [reply services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java:80]: Yeah... test code is very repetitive... it's nice to DRY it up a bit :D ...

File: internal/venice-common/src/test/java/com/linkedin/venice/meta/TestZKStore.java (hunk 1 above)
Comment (FelixGV) [reply internal/venice-common/src/test/java/com/linkedin/venice/meta/TestZKStore.java:402]: Included that in the latest commit message (in auto-merge).

File: services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java (hunk 4 above)
Comment (FelixGV) [reply services/venice-router/src/test/java/com/linkedin/venice/router/throttle/RouterRequestThrottlingTest.java]: This is now addressed in the latest commit, but needs another ship it...

**Review by KaiSernLim — APPROVED** [review]

**Pull request merged**